pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go

```diff
@@ -147,50 +148,60 @@ func (admitter *VMsAdmitter) authorizeVirtualMachineSpec(ar *admissionv1.Admissi
 	var causes []metav1.StatusCause
 
 	for idx, dataVolume := range vm.Spec.DataVolumeTemplates {
-		pvcSource := dataVolume.Spec.Source.PVC
-		if pvcSource != nil {
-			sourceNamespace := pvcSource.Namespace
-			if sourceNamespace == "" {
-				if vm.Namespace != "" {
-					sourceNamespace = vm.Namespace
-				} else {
-					sourceNamespace = ar.Namespace
-				}
-			}
+		cloneSource, err := typesutil.GetCloneSource(context.TODO(), admitter.virtClient, vm, &dataVolume.Spec)
+		if err != nil {
+			causes = append(causes, metav1.StatusCause{
+				Type:    metav1.CauseTypeUnexpectedServerResponse,
+				Message: err.Error(),
+				Field:   k8sfield.NewPath("spec", "dataVolumeTemplates").Index(idx).String(),
+			})
 
-			if sourceNamespace == "" || pvcSource.Name == "" {
-				causes = append(causes, metav1.StatusCause{
-					Type:    metav1.CauseTypeFieldValueNotFound,
-					Message: fmt.Sprintf("Clone source %s/%s invalid", sourceNamespace, pvcSource.Name),
-					Field:   k8sfield.NewPath("spec", "dataVolumeTemplates").Index(idx).String(),
-				})
-			} else {
-				targetNamespace := vm.Namespace
-				if targetNamespace == "" {
-					targetNamespace = ar.Namespace
-				}
+			continue
+		}
 
-				serviceAccount := "default"
-				for _, vol := range vm.Spec.Template.Spec.Volumes {
-					if vol.ServiceAccount != nil {
-						serviceAccount = vol.ServiceAccount.ServiceAccountName
-					}
-				}
+		if cloneSource == nil {
+			continue
+		}
 
-				allowed, message, err := admitter.cloneAuthFunc(sourceNamespace, pvcSource.Name, targetNamespace, serviceAccount)
-				if err != nil {
-					return nil, err
-				}
+		if cloneSource.Namespace == "" {
+			cloneSource.Namespace = ar.Namespace
```

Comment:
Can ar.Namespace be ""? I don't think so, and if it cannot be blank, then the check on line 170 can skip the cloneSource.Name == "" part.

Reply:
Technically I think you are correct.  Will update if there are any other changes requested